opm/simulators/linalg/bda/CPR.cpp

```diff
@@ -0,0 +1,515 @@
+    diagIndices.resize(num_levels);
+
+    Amatrices.reserve(num_levels);
+    Pmatrices.reserve(num_levels - 1);  // coarsest level does not need one
+    Rmatrices.reserve(num_levels - 1);
+    invDiags.reserve(num_levels);
+
+    Amatrices.clear();
+    invDiags.clear();
+
+    // matrixIter.dereference() returns MatrixAdapter
+    // matrixIter.dereference().getmat() returns BCRSMat
+    DuneAmg::ParallelMatrixHierarchy::ConstIterator matrixIter = matrixHierarchy.finest();
+    for(int level = 0; level < num_levels; ++matrixIter, ++level) {
+        const auto& A = matrixIter.dereference().getmat();
+        level_sizes[level] = A.N();
+        diagIndices[level].reserve(A.N());
+
+        // extract matrix A
+        Amatrices.emplace_back(A.N(), A.nonzeroes());
+        // contiguous copy is not possible
+        // std::copy(&(A[0][0][0][0]), &(A[0][0][0][0]) + A.nonzeroes(), Amatrices.back().nnzValues.data());
+        // also update diagonal indices if needed, level 0 is already filled in create_preconditioner()
+        int nnz_idx = 0;
+        const bool fillDiagIndices = diagIndices[level].empty();
+        for (typename DuneMat::const_iterator r = A.begin(); r != A.end(); ++r) {
+            for (auto c = r->begin(); c != r->end(); ++c) {
+                Amatrices.back().nnzValues[nnz_idx] = A[r.index()][c.index()];
+                if (fillDiagIndices && r.index() == c.index()) {
+                    diagIndices[level].emplace_back(nnz_idx);
+                }
+                nnz_idx++;
+            }
+        }
+
+        Opm::BdaBridge<DuneMat, DuneVec, 1>::getSparsityPattern(A, Amatrices.back().rowPointers, Amatrices.back().colIndices);
+
+        // compute inverse diagonal values for current level
+        invDiags.emplace_back(A.N());
+        for (unsigned int row = 0; row < A.N(); ++row) {
+            invDiags.back()[row] = 1 / Amatrices.back().nnzValues[diagIndices[level][row]];
+        }
+    }
+}
+
+
+template <unsigned int block_size>
+void CPR<block_size>::analyzeAggregateMaps() {
+
+    Pmatrices.clear();
+    Rmatrices.clear();
+
+    const DuneAmg::AggregatesMapList& aggregatesMaps = dune_amg->aggregatesMaps();
+
+    DuneAmg::AggregatesMapList::const_iterator mapIter = aggregatesMaps.begin();
+    for(int level = 0; level < num_levels - 1; ++mapIter, ++level) {
+        DuneAmg::AggregatesMap *map = *mapIter;
+
+        // get indices for each row of P and R
+        std::vector<std::set<long int> > indicesP(level_sizes[level]);
+        std::vector<std::set<long int> > indicesR(level_sizes[level+1]);
+        using AggregateIterator = DuneAmg::AggregatesMap::const_iterator;
+        for(AggregateIterator ai = map->begin(); ai != map->end(); ++ai){
+            if (*ai != DuneAmg::AggregatesMap::ISOLATED) {
+                const long int diff = ai - map->begin();
+                indicesP[diff].insert(*ai);
+                indicesR[*ai].insert(diff);
+            }
+        }
```

Comment:
I think this is a bit of overkill (even if you insist on using matrices for both P and R). Unless a node isolated the corresponding row of P should only have one entry otherwise none. Hence the set will only have one entry or none. That could be determined withour set (and its complexity), using e.g. a vector with either 1 or 0.

For R you need to know the number of dofs per aggregate. Here you could simply count these (e.g. in a vector containing the size for each coarse grid dof) and then loop over the aggregates map again to determine the column index when creating the sparsity pattern.

Complexity of iterating a second time over the aggregates map should not be worse than over two vectors of sets. (Famous last words...)

Reply:
I managed to remove the P matrices, and replaced them with a vector of just colIndices and a new kernel.
R is also filled differently. I'm not sure what you mean by iterating the aggregates map twice, I only see one.